### PR TITLE
refactor: avoid `RwError` in schema registry client

### PR DESCRIPTION
**Pull request by xiangjinwu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

* `InvalidOptionError` is returned by `handle_sr_list`, `schema_registry::Client::new`, `get_subject_by_strategy` and avro/protobuf sink. It can be converted into `SchemaFetchError` (for sink) and `RwError` (for source).
* `ConcurrentRequestError` (multiple `RequestError`s) is returned by schema registry API calls and can be converted into `SchemaFetchError` and `RwError`.
* `SchemaFetchError::YetToMigrate` for protobuf sink will be made into `InvalidOption` / `Request` / `SchemaCompile` in the next PR.
* `extract_schema_id` and `WireFormatError` is only used in source (`RwError`).

Refer to https://gist.github.com/xiangjinwu/3f25addc675ed56721eb9fff7a742b96 for the differences between old and new error messages in all 8 error bubbling paths. They are not added to `e2e_test/error_ui`.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/connector/src/schema/schema_registry/client.rs [hunk 1]
```diff
@@ -61,29 +61,44 @@ pub struct Client {
...
                 valid_urls
             );
         }
 
-        let inner = reqwest::Client::builder().build().map_err(|e| {
-            RwError::from(ProtocolError(format!("build reqwest client failed {}", e)))
-        })?;
+        let inner = reqwest::Client::builder()
+            .build()
+            .map_err(|e| invalid_option_error!("build reqwest client failed: {}", e.as_report()))?;
```
Comment (xiangjinwu) [inline src/connector/src/schema/schema_registry/client.rs:101]: Shall `invalid_option_error!` macro accept `reqwest::Error` (or `Into<BoxedError>`) as `#[source]` field?

File: src/connector/src/schema/schema_registry/util.rs [hunk 2]
```diff
@@ -34,10 +33,7 @@ pub fn handle_sr_list(addr: &str) -> Result<Vec<Url>> {
         }
     }
     if urls.is_empty() {
-        return Err(RwError::from(ProtocolError(format!(
-            "no valid url provided, got {:?}",
-            errs
-        ))));
+        bail_invalid_option_error!("no valid url provided, errs: {errs:?}");
```
Comment (xiangjinwu) [inline src/connector/src/schema/schema_registry/util.rs:36]: `Vec<ParseError>` cannot be returned as `std::error::Error::source`

File: src/connector/src/schema/schema_registry/client.rs (hunk 1 above)
Comment (BugenZhao) [reply src/connector/src/schema/schema_registry/client.rs:101]: That would be better. The source error could be displayed in a multi-line manner in pretty mode.

File: src/connector/src/schema/mod.rs [hunk 3]
```diff
@@ -22,5 +22,27 @@ const SCHEMA_LOCATION_KEY: &str = "schema.location";
...
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum SchemaFetchError {
+    #[error(transparent)]
+    InvalidOption(#[from] InvalidOptionError),
+    #[error(transparent)]
+    Request(#[from] schema_registry::ConcurrentRequestError),
+    #[error("schema compilation error")]
```
Comment (BugenZhao) [inline src/connector/src/schema/mod.rs:44]: For now, we may still include the message of the source error here, as #13763 is still not completed and an upper error may directly format with `.to_string()`.

File: src/connector/src/schema/schema_registry/client.rs [hunk 4]
```diff
@@ -61,29 +61,44 @@ pub struct Client {
     password: Option<String>,
 }
 
+#[derive(Debug, thiserror::Error)]
+#[error("all request confluent registry all timeout, {context}\n{}", errs.iter().map(|e| format!("\t{}", e.as_report())).join("\n"))]
```
Comment (BugenZhao) [inline src/connector/src/schema/schema_registry/client.rs:65]: I guess a `MultiError` is really a good to have. 😂

File: src/connector/src/schema/mod.rs [hunk 5]
```diff
@@ -22,5 +22,27 @@ const SCHEMA_LOCATION_KEY: &str = "schema.location";
...
+    SchemaCompile(#[source] risingwave_common::error::BoxedError),
```
[the remark this answers is not in the document]
Comment (tabVersion) [reply src/connector/src/schema/mod.rs:45]: BoxedError is enough here I think.

**Review by tabVersion — APPROVED** [review]

File: src/connector/src/schema/mod.rs [hunk 6]
```diff
@@ -22,5 +22,27 @@ const SCHEMA_LOCATION_KEY: &str = "schema.location";
...
+#[derive(Debug, thiserror::Error)]
+pub enum SchemaFetchError {
+    #[error(transparent)]
+    InvalidOption(#[from] InvalidOptionError),
+    #[error(transparent)]
+    Request(#[from] schema_registry::ConcurrentRequestError),
+    #[error("schema compilation error: {0}")]
+    SchemaCompile(#[source] risingwave_common::error::BoxedError),
+    #[error(transparent)]
+    YetToMigrate(risingwave_common::error::RwError),
```
Comment (BugenZhao) [inline src/connector/src/schema/mod.rs:47]:
```suggestion
    SchemaCompile(#[source] #[backtrace] risingwave_common::error::BoxedError),
    #[error(transparent)]
    YetToMigrate(#[backtrace] risingwave_common::error::RwError),
```

**Review by BugenZhao — APPROVED** [review]
Rest LGTM

File: src/connector/src/schema/schema_registry/client.rs (hunk 1 above)
Comment (xiangjinwu) [reply src/connector/src/schema/schema_registry/client.rs]: Updated to `unwrap` for this specific case. The call is not affected by any input.

**Pull request merged**